Source/Core/UICommon/GameFileCache.cpp

```diff
@@ -0,0 +1,235 @@
+{
+  // Copy game paths into a set, except ones that match DiscIO::ShouldHideFromGameList.
+  // TODO: Prevent DoFileSearch from looking inside /files/ directories of DirectoryBlobs at all?
+  // TODO: Make DoFileSearch support filter predicates so we don't have remove things afterwards?
+  std::unordered_set<std::string> game_paths;
+  game_paths.reserve(all_game_paths.size());
+  for (const std::string& path : all_game_paths)
+  {
+    if (!DiscIO::ShouldHideFromGameList(path))
+      game_paths.insert(path);
+  }
+
+  bool cache_changed = false;
+
+  // Delete paths that aren't in game_paths from m_cached_files,
+  // while simultaneously deleting paths that aren't in m_cached_files from game_paths.
+  // For the sake of speed, we don't care about maintaining the order of m_cached_files.
+  {
+    auto it = m_cached_files.begin();
+    auto end = m_cached_files.end();
+    while (it != end)
+    {
+      if (game_paths.erase((*it)->GetFilePath()))
+      {
+        ++it;
+      }
+      else
+      {
+        cache_changed = true;
+        --end;
+        *it = std::move(*end);
+      }
+    }
+    m_cached_files.erase(it, m_cached_files.end());
+  }
+
+  // Now that the previous loop has run, game_paths only contains paths that
+  // aren't in m_cached_files, so we simply add all of them to m_cached_files.
+  for (const auto& path : game_paths)
+  {
+    auto file = std::make_shared<GameFile>(path);
+    if (file->IsValid())
+    {
+      cache_changed = true;
+      m_cached_files.push_back(std::move(file));
+    }
+  }
+
+  return cache_changed;
+}
+
+bool GameFileCache::UpdateAdditionalMetadata(const Core::TitleDatabase& title_database)
+{
+  bool cache_changed = false;
+
+  for (auto& file : m_cached_files)
+    cache_changed |= UpdateAdditionalMetadata(&file, title_database);
+
+  return cache_changed;
+}
+
+bool GameFileCache::UpdateAdditionalMetadata(std::shared_ptr<GameFile>* game_file,
+                                             const Core::TitleDatabase& title_database)
+{
+  const bool emu_state_changed = (*game_file)->EmuStateChanged();
+  const bool banner_changed = (*game_file)->BannerChanged();
+  const bool custom_title_changed = (*game_file)->CustomNameChanged(title_database);
+  if (!emu_state_changed && !banner_changed && !custom_title_changed)
+    return false;
+
+  // If a cached file needs an update, apply the updates to a copy and delete the original.
+  // This makes the usage of cached files in other threads safe.
+
+  std::shared_ptr<GameFile> copy = std::make_shared<GameFile>(**game_file);
+  if (emu_state_changed)
+    copy->EmuStateCommit();
+  if (banner_changed)
+    copy->BannerCommit();
+  if (custom_title_changed)
+    copy->CustomNameCommit();
+  *game_file = std::move(copy);
+
+  return true;
+}
+
+bool GameFileCache::Load()
+{
+  return SyncCacheFile(false);
+}
+
+bool GameFileCache::Save()
+{
+  return SyncCacheFile(true);
+}
+
+bool GameFileCache::SyncCacheFile(bool write)
```

Comment:
Might be a good idea to avoid the name "write", as that is an actual function name?

Reply:
Would `save` be better?

Reply:
Sounds good to me